### DISCUSSION TITLE
Delaunay triangulation 3D with CGAL

### DISCUSSION THREAD

**Discussion by ReeNine** [CGAL/cgal#6140] [Q&A] [answered]
I am trying to use CGAL to do some Delaunay triangulation. I used one of the CGAL samples to compute a triangulation which includes a height field attribute.

The problem I have having is that I have no idea how to get the resulting triangulation. I figured out how to get the face_iterator, but I don't know what to do from there. What I'm hoping to get is an index into the point array for each of the 3 points on each triangle.

Im just wanna get smth like: Triang: triangle1 = (1,2,3) triangle2 = (3,4,1) where numbers its a vertices HELP PLEASE!!!

This is my code:

```
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>
#include <CGAL/Delaunay_triangulation_3.h>
#include <CGAL/Random.h>
#include <CGAL/Triangulation_3.h>
#include <CGAL/Delaunay_triangulation_cell_base_3.h>
#include <CGAL/Triangulation_vertex_base_with_info_3.h>


#include <vector>
#include <cassert>
#include <iostream>
#include <fstream>

using namespace std;

typedef CGAL::Exact_predicates_inexact_constructions_kernel     K;
typedef CGAL::Delaunay_triangulation_3<K, CGAL::Fast_location>  Delaunay;
typedef Delaunay::Point                                         Point;
typedef Delaunay::Vertex_handle                                 Vertex_handle;
typedef Delaunay::Cell_handle                                   Cell_handle;
typedef Delaunay::All_cells_iterator                            Cells_iterator;
typedef Delaunay::Vertex                                        Vertex;

typedef CGAL::Triangulation_vertex_base_with_info_3<unsigned, K>    Vb;
typedef CGAL::Delaunay_triangulation_cell_base_3<K>                 Cb;
typedef CGAL::Triangulation_data_structure_3<Vb, Cb>                Tds;

typedef Delaunay::Finite_vertices_iterator                      Finite_vertices_iterator;
typedef Delaunay::Finite_edges_iterator                         Finite_edges_iterator;
typedef Delaunay::Finite_facets_iterator                        Finite_facets_iterator;
typedef Delaunay::Finite_cells_iterator Finite_cells_iterator;
typedef Delaunay::Simplex        Simplex;
typedef Delaunay::Locate_type    Locate_type;




int main()
{
    // generating points on a grid.
    std::vector<Point> P;

    for (int z = 0; z < 3; z++)
        for (int y = 0; y < 3; y++)
            for (int x = 0; x < 3; x++)
                P.push_back(Point(x, y, z));

    // building their Delaunay triangulation.
    Delaunay T(P.begin(), P.end());
    //assert(T.number_of_vertices() == 8000);

    for (?????????)
    {
           ????????????????????????
    }

    std::ofstream oFileT("output", std::ios::out);
    // writing file output;
    oFileT << T;
    cout << T.is_valid() << endl;
    return 0;
}
```

**Comment by sloriot** [comment, marked as answer]
As documented [here](https://doc.cgal.org/latest/TDS_3/classTriangulationDataStructure__3.html#ad6a20b45e66dfb690bfcdb8438e9fcae)  a face is a pair containing a cell and the id in the cell of the vertex opposite to that face. So you can access vertices of the face using:
```
for (Delaunay::Finite_facets_iterator  fit=T.finite_facets_begin(), end=T.finite_facets_end(); fit!=end; ++fit)
{
  Delaunay::Facet f = *fit;
  Delaunay::Vertex_handle vh0 = f.first->vertex( T.vertex_triple_index(f.second, 0) );
  Delaunay::Vertex_handle vh1 = f.first->vertex( T.vertex_triple_index(f.second, 1) );
  Delaunay::Vertex_handle vh2 = f.first->vertex( T.vertex_triple_index(f.second, 2) );
}
```
This will give you handles to vertices of the facet.

If you need input ids, then you have to use something like in [this example](https://github.com/CGAL/cgal/blob/master/Triangulation_3/examples/Triangulation_3/info_insert_with_transform_iterator.cpp). To access the id from a vertex handle, use `vh->info()`.

**Reply by ReeNine** [reply]
TY very much. Do you by any chance know how to get only the triangulation that will be outside?

**Reply by MagicalNamNam** [reply]
I also want to know.Have you sloved it already?(●'◡'●)

**Reply by afabri** [reply]
Your problem statement is not clear.

**Reply by MagicalNamNam** [reply]
Sorry, I want to reconstruct the mesh from the point cloud I got but do not want to use methods like marching cube. So I tried to use the delaunay triangulation then I found the result mesh containing many triangles are inside the mesh.  How can I get the surface triangles?

**Reply by sloriot** [reply]
check https://doc.cgal.org/latest/Manual/tuto_reconstruction.html

**Reply by afabri** [reply]
Also when you have a terrain, then you are probably better off with a 2D triangulation.

**Reply by MagicalNamNam** [reply]
Thank you. I have tried the Advanced Font method, the watertight mesh may appear some holes and the non-watertight mesh's hole may be covered by some slim faces(though the hole is big). I will try more.
